### ISSUE TITLE
[0.7.1] GIL test failed with Python3.5

### ISSUE THREAD

**Issue by sagitter** [uweschmitt/autowrap#36]
Hi all,

GIL test is not compiled with Python3.5:

```
gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -march=armv7-a -mfpu=vfpv3-d16 -mfloat-abi=hard -D_GNU_SOURCE -fPIC -fwrapv -I/usr/include/boost -fPIC -I/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/autowrap/data_files/boost -I/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/autowrap/data_files -I/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/tests/test_files -I/usr/include/python3.5m -c gil_testing_wrapper.cpp -o build/temp.linux-armv7l-3.5/gil_testing_wrapper.o -Wno-unused-but-set-variable
In file included from gil_testing_wrapper.cpp:260:0:
/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/tests/test_files/gil_testing.hpp: In member function 'void GilTesting::do_something(const char*)':
/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/tests/test_files/gil_testing.hpp:14:75: error: '_PyThreadState_Current' was not declared in this scope
         PyThreadState * tstate = (PyThreadState*)_Py_atomic_load_relaxed(&_PyThreadState_Current);
                                                                           ^
/builddir/build/BUILD/python3-autowrap-0.7.0-1.20151124gitd5e233.fc24/tests/test_files/gil_testing.hpp:14:97: error: '_Py_atomic_load_relaxed' was not declared in this scope
         PyThreadState * tstate = (PyThreadState*)_Py_atomic_load_relaxed(&_PyThreadState_Current);
                                                                                                 ^
error: command 'gcc' failed with exit status 1
FAIL
```

Full log: https://kojipkgs.fedoraproject.org//work/tasks/3203/11993203/build.log

**Comment by sagitter** [comment]
@uweschmitt 

This error persists with ```autowrap-0.7.1```
https://kojipkgs.fedoraproject.org//work/tasks/5085/12015085/build.log

**Comment by hroest** [comment]
this may be related to https://bugs.python.org/issue23644

**Comment by hroest** [comment]
I tried to do this
```
-        PyThreadState * tstate = (PyThreadState*)_Py_atomic_load_relaxed(&_PyThreadState_Current);
+        PyThreadState * tstate = PyThreadState_GET();
```
in `tests/test_files/gil_testing.hpp` as the Python headers indicated that this should be an alias for Python 3.3 and 3.4 while it is more complicated for 3.5. Apparently in 3.5 it expects to have access to C++11 functions and thus uses a different macro to get the thread state

 and even though this compiles, it leads to a segfault:

```
Fatal Python error: PyThreadState_Get: no current thread
Segmentation fault (core dumped)

```

**Comment by hroest** [comment]
seems like in Python 3.4 and higher there is an easier way to check for the GIL

http://stackoverflow.com/questions/11366556/how-can-i-check-whether-a-thread-currently-holds-the-gil

**Comment by hroest** [comment]
for a fix, see #38

**Comment by hroest** [comment]
@uweschmitt can we close this?
